### PR TITLE
realmd: Don't directly manipulate the DOM

### PR DESCRIPTION
**Pull request by mvollmer**
Both the realmd code and the systemd code would try to to set the
"disabled" attribute of the #system_information_hostname_button
element, and would race each other.

To get better decoupling, the realmd code has been changed to only
export the necessary information to construct the buttons (instead of
modifying the DOM directly with jQuery), and the systemd code now uses
that to render them with React.

As part of that the realmd code hase been moved into pkg/systemd since
that is the only user.

### CODE REVIEW

**Comment by mvollmer** [comment]
The automated tests need updating and I need to do some proper manual testing.

File: pkg/systemd/overview-cards/realmd-operation.js [hunk 1]
```diff
@@ -564,14 +562,20 @@ function instance(realmd, mode, realm, button) {
...
-    element.append(link);
-    var hostname_link = $("#system_information_hostname_button");
-    var hostname_tooltip = $("#system_information_hostname_tooltip");
+    var self = {
+        button_text: "DOMAIN!",
+        button_tooltip: "Not for you",
+        button_disabled: true,
+        hostname_button_tooltip: "Nope",
+        hostname_button_disabled: true,
+
```
Comment (mvollmer) [inline pkg/systemd/overview-cards/realmd-operation.js]: These default values are silly and need to be removed, of course.

File: pkg/systemd/overview-cards/configurationCard.jsx [hunk 2]
```diff
@@ -94,6 +97,10 @@ export class ConfigurationCard extends React.Component {
         });
 
         $("#system_information_ssh_keys").on("hide.bs.modal", () => this.host_keys_hide());
+
+        this.realmd.addEventListener("changed", () => {
+            this.setState({});
+        });
```
Comment (martinpitt) [inline pkg/systemd/overview-cards/configurationCard.jsx]: Please no function blocks for simple expressions.

File: pkg/systemd/README-realmd.md [hunk 3]
```diff
@@ -1,3 +1,5 @@
+!! THIS IS A BIT OUT OF DATE BUT MIGHT STILL BE USEFUL !!
```
Comment (martinpitt) [inline pkg/systemd/README-realmd.md]: Should actually be okay except that the image is now "services", and one needs to call /run-freeipa like test/verify/check-realms does.

File: test/verify/check-realms [hunk 4]
```diff
@@ -95,7 +95,9 @@ class TestRealms(MachineCase):
         if self.machine.image in ["rhel-8-1-distropkg"]:
             self.domain_sel = "#system-info-domain a"
         else:
-            self.domain_sel = "#system-info-domain button"
+            self.domain_sel = "#system_information_domain_button"
+
+        self.machine.upload([ "/home/mvo/work/cockpit/dist/systemd" ], "/usr/share/cockpit" )
```
Comment (martinpitt) [inline test/verify/check-realms]: Please remember to take this out for the final commit :-)

File: tools/cockpit.spec [hunk 5]
```diff
@@ -352,7 +349,6 @@ Requires: shadow-utils
 Requires: grep
 Requires: libpwquality
 Requires: /usr/bin/date
-Provides: cockpit-realmd = %{version}-%{release}
```
Comment (martinpitt) [inline tools/cockpit.spec]: This is an independent change, and I think it's ok to leave this for now.

File: tools/debian/control [hunk 6]
```diff
@@ -159,8 +159,7 @@ Depends: ${misc:Depends},
          libpwquality-tools,
          openssl,
 Recommends: policykit-1
-Provides: cockpit-realmd,
```
Comment (martinpitt) [inline tools/debian/control]: Indeed cockpit-realmd has never been in Debian/Ubuntu as an explicit package, but again it's an unrelated change. Separate commit for cleaning it up is fine if you want to keep it.

File: pkg/systemd/overview-cards/configurationCard.jsx [hunk 7]
```diff
@@ -321,14 +328,27 @@ export class ConfigurationCard extends React.Component {
     }
 
     render() {
+        // We use a Privileged component for its ability to
+        // conditionally show a tooltip, even when the button is not
+        // actually disabled, so the "allowed" property really means
+        // "does not have tooltip" here.
+
+        const hostname_tooltip = (this.permission.allowed == false
```
Comment (martinpitt) [inline pkg/systemd/overview-cards/configurationCard.jsx]: Please use `===` for `false` comparison.

File: pkg/systemd/overview-cards/configurationCard.jsx [hunk 8]
```diff
@@ -321,14 +328,27 @@ export class ConfigurationCard extends React.Component {
...
+            ? cockpit.format(_("The user $0 is not permitted to modify hostnames"),
+                             this.permission.user ? this.permission.user.name : '')
+            : this.realmd.hostname_button_tooltip);
+        const hostname_disabled = this.permission.allowed == false || this.realmd.hostname_button_disabled;
+
         const hostname_button = (
-            <PrivilegedButton variant="link" buttonId="system_information_hostname_button"
```
Comment (martinpitt) [inline pkg/systemd/overview-cards/configurationCard.jsx]: Perhaps we should teach PrivilegedButton to have an additional disabled reason -- I think on the Storage page you also have that case, right? I'm happy to help out with this, given that I wrote this component.

File: tools/cockpit.spec (hunk 5 above)
Comment (mvollmer) [reply tools/cockpit.spec]: Yes, I was wondering about that.  My reasoning for removing the `Provides` is that installing cockpit-realmd probably is meant to give you `/usr/share/cockpit/realmd`, but this this directory no longer exists...

File: pkg/systemd/overview-cards/configurationCard.jsx (hunk 8 above)
Comment (mvollmer) [reply pkg/systemd/overview-cards/configurationCard.jsx]: We also need the "tooltip but enabled" case here, so using long-hand with both Privileged and Button is probably best.

File: tools/cockpit.spec (hunk 5 above)
Comment (martinpitt) [reply tools/cockpit.spec]: Right, that was a thinko of mine yesterday. With this PR, one can't install cockpit-realmd independently any more (conceptually).

File: tools/debian/control (hunk 6 above)
Comment (martinpitt) [reply tools/debian/control]: Nevermind, same reason as for the .spec file, my comment was bogus.

File: pkg/systemd/overview-cards/configurationCard.jsx (hunk 7 above)
Comment (mvollmer) [reply pkg/systemd/overview-cards/configurationCard.jsx]: Done, thanks.

File: tools/debian/control (hunk 6 above)
Comment (mvollmer) [reply tools/debian/control]: Removed this change.

File: tools/cockpit.spec (hunk 5 above)
Comment (mvollmer) [reply tools/cockpit.spec]: Removed this change.

File: pkg/systemd/overview-cards/realmd-operation.js (hunk 1 above)
Comment (mvollmer) [reply pkg/systemd/overview-cards/realmd-operation.js]: Done.

File: pkg/systemd/overview-cards/configurationCard.jsx (hunk 2 above)
Comment (mvollmer) [reply pkg/systemd/overview-cards/configurationCard.jsx]: Okay. :-)  (I like to use statement blocks for side-effects, and expressions for side-effect free things.)

File: test/verify/check-realms (hunk 4 above)
Comment (mvollmer) [reply test/verify/check-realms]: Done.

File: tools/cockpit.spec (hunk 5 above)
Comment (mvollmer) [reply tools/cockpit.spec]: Oh, misunderstanding.  I put the change back.

File: tools/debian/control (hunk 6 above)
Comment (mvollmer) [reply tools/debian/control]: Oh, misunderstanding. I put the change back.

**Comment by mvollmer** [comment]
There was this failure https://209.132.184.41:8493/logs/pull-13369-20200109-134242-3539090f-cockpit-project-cockpit--debian-stable/log.html#262-2. It doesn't seem to be new, but I'll try to gather some information anyway.

File: pkg/systemd/overview-cards/configurationCard.jsx [hunk 9]
```diff
@@ -339,6 +357,23 @@ export class ConfigurationCard extends React.Component {
...
+            : this.realmd.button_tooltip);
+        const domain_disabled = this.permission.allowed === false || this.realmd.button_disabled;
+
+        const domain_button = (
+            <Privileged allowed={ !domain_tooltip }
+                        tooltipId="system_information_domain_tooltip"
+                        excuse={ domain_tooltip }>
+                <Button id="system_information_domain_button" variant="link"
+                        onClick={ () => this.realmd.clicked() }
+                        isInline isDisabled={ domain_disabled } aria-label="join domain">
```
Comment (martinpitt) [inline pkg/systemd/overview-cards/configurationCard.jsx]: Do we really need the aria-label here? This is just a standard button, and the actual text is so much more correct. I. e. when you are already joined, then "join domain" is wrong, and it's also not translated. If we need it, could it just be `this.realmd.button_text` instead?

File: test/verify/check-realms [hunk 10]
```diff
@@ -781,7 +781,20 @@ class TestPackageInstall(packagelib.PackageCase):
...
+
+    def checkTooltip(self, text):
+        b = self.browser
+        if self.machine.image in ["rhel-8-1-distropkg"]:
+            b.wait_present("#system-info-domain [data-original-title]")
+            b.mouse("#system-info-domain span", "mouseover")
+            b.wait_in_text(".tooltip-inner", text)
+            b.mouse("#system-info-domain span", "mouseout")
+        else:
+            b.wait_present("#system_information_domain_tooltip[data-test-has-tooltip]")
```
Comment (martinpitt) [inline test/verify/check-realms]: This is the counterpart to the attribute definition above. What does that line do? AFAICS, just dropping it should have exactly the same result -- it waits  for `system_information_domain_tooltip`, hovers it, and then waits for the actual tooltip to pop up and checks its text. Why is that not sufficient?

Comment (mvollmer) [reply test/verify/check-realms]: The `#system_information_domain_tooltip` is always present, even if there is no tooltip.  We might indeed change so that it only appears when there is a tooltip.  (Fingers crossed that this doesn't screw with the layout.)

**Comment by mvollmer** [comment]
> There was this failure https://209.132.184.41:8493/logs/pull-13369-20200109-134242-3539090f-cockpit-project-cockpit--debian-stable/log.html#262-2. It doesn't seem to be new, but I'll try to gather some information anyway.

Fix in #13377.

File: pkg/lib/cockpit-components-privileged.jsx [hunk 11]
```diff
@@ -34,7 +34,7 @@ import cockpit from "cockpit";
  */
 export function Privileged({ excuse, allowed, placement, tooltipId, children }) {
     // wrap into extra <span> so that a disabled child keeps the tooltip working
-    let contents = <span id={tooltipId}>{ children }</span>;
+    let contents = <span id={tooltipId} data-test-has-tooltip={allowed ? null : ""}>{ children }</span>;
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/lib/cockpit-components-privileged.jsx]: I think the `div.tooltip` only appears once the tooltip is actually open.  We also need to wait until the `Overlaytrigger` has been put into the DOM so that we can send the `mouseover` that will open the tooltip.

File: test/verify/check-realms (hunk 10 above)
Comment (martinpitt) [reply test/verify/check-realms]: Right, that's the "hook" for attaching a notification to, and we need it for the hovering. It's fine to always be present. My question was, why do we need to wait for this attribute in the first place? It should be equivalent to waiting for `div.tooltip` to appear.

File: pkg/systemd/overview-cards/configurationCard.jsx (hunk 9 above)
Comment (mvollmer) [reply pkg/systemd/overview-cards/configurationCard.jsx]: It was in the original code, so... *shrug*

Comment (martinpitt) [reply pkg/systemd/overview-cards/configurationCard.jsx]: Ah, I didn't see it in the diff. Nevermind then.

File: pkg/lib/cockpit-components-privileged.jsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/lib/cockpit-components-privileged.jsx]: We don't do weird things in the other tests that check tooltips. `b.mouse()` already waits for the element to become present and visible.

File: test/verify/check-realms (hunk 10 above)
Comment (mvollmer) [reply test/verify/check-realms]: Yes, removing the `<span>` unfortunately messes with the hostname change button.  So what about leaving the `<span>` in but only giving it the `id` when it actually has a tooltip?  I have done that.

**Comment by mvollmer** [comment]
> Unresolving, as this shouldn't be in the final commit. Please change s/ipa/services/ instead.

Done.

File: pkg/lib/cockpit-components-privileged.jsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/lib/cockpit-components-privileged.jsx]: In the other cases, the test is in a state where we know that the tooltip is in the DOM.  For example, after being joined to a domain, the hostname edit button is expected to have a tooltip. We don't need to wait for it.

In the TestInstall cases, there is no explicit action that the tests do that would guarantee that the tooltip is there.  It really appears asynchronously and we need to wait for it. The old code also did this, with             
```
b.wait_present("#system-info-domain [data-original-title]")
```

[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/lib/cockpit-components-privileged.jsx]: Ah, you mean this doesn't wait for the tooltip, but it waits for the page to get into a state where the action of hovering *will cause* the tooltip to appear. That's the bit that I was missing, sorry.

File: test/verify/check-realms (hunk 10 above)
Comment (martinpitt) [reply test/verify/check-realms]: Right, you need the span so that the tooltip gets attached to a non-disabled element; tooltips on disabled elements don't work. This still looks funny, but I think I understand the reason now. Thanks!

File: test/verify/check-realms [hunk 12]
```diff
@@ -878,10 +875,12 @@ class TestPackageInstall(packagelib.PackageCase):
...
-        b.mouse("#system-info-domain span", "mouseover")
-        time.sleep(2)
-        self.assertFalse(b.is_present(".tooltip-inner"))
-
+        if self.machine.image in ["rhel-8-1-distropkg"]:
+            b.mouse("#system-info-domain span", "mouseover")
+            time.sleep(2)
+            self.assertFalse(b.is_present(".tooltip-inner"))
+        else:
+            b.wait_attr("#system_information_domain_tooltip", "data-test-has-tooltip", None)
```
Comment (martinpitt) [inline test/verify/check-realms]: That element now doesn't exist any more if there's no tooltip. It should become  `bwait_not_present(`#system_information_domain_tooltip`).

Comment (mvollmer) [reply test/verify/check-realms]: Ouch, well spotted.

File: pkg/systemd/README-realmd.md (hunk 3 above)
Comment (mvollmer) [reply pkg/systemd/README-realmd.md]: Fixed
